### ISSUE TITLE
DOC/DX: direct to devdocs from dev section of released docs

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#19699]
### Issue with current documentation:

People (especially new contributors) can be led astray if they end up on an outdated version of the development docs (and don't realise it).

### Idea or request for content:

- Add a note to the dev sections of released docs to direct people to the devdocs (as this is almost always what they want).

An alternative would be to automatically redirect, but I think it makes sense to have the HTML pages available for old versions, even if their use would be rare.

### Additional context (e.g. screenshots, GIFs)

- Would probably be nice to get the version switcher fixed before this (gh-18031).

**Comment by lucascolley** [comment]
@tylerjereddy how could we go about this? A PR to maintenance branch + backports + a PR every release, or is there a more sophisticated way to hide notes from the devdocs?

(Sidenote: I vaguely remember reading something about merging the two docs websites into one, I don't know how speculative that was though)

**Comment by tylerjereddy** [comment]
I think @tupui had thoughts on this kind of stuff. I'm not fully sure what you mean just yet though:

> led astray if they end up on an outdated version of the development docs

You mean landing on a stable release that is older than current stable or? There is the red color for example on an older release:

![image](https://github.com/scipy/scipy/assets/7903078/bc155180-d4cb-4f54-bbc7-8d7387444027)

Although it currently shows up as red for me on `1.11.4` as well, but there's also the version switcher issue. There may be a few different problems that need to be solved. Normally wouldn't want to backport just for a warning banner or something, sometimes we manually apply stuff right on the docs server but that's also risky without peer review, etc.

**Comment by tylerjereddy** [comment]
At least for me the dev link works from a stable release, no?

**Comment by lucascolley** [comment]
Sorry, I think the message was lost in the terminology here, let me elaborate.

This PR is just about the `/dev` section of the docs (https://scipy.github.io/devdocs/dev/index.html).

The specific problem is a user landing on e.g. https://docs.scipy.org/doc/scipy-1.11.4/dev/index.html (any released version, stable or old - I imagine the stable case is not uncommon). Sometimes changes are made to the `/dev` section between releases, and it can be confusing if new contributors are reading outdated info (especially when there are changes to build instructions[^1]). The problem only gets worse the older the version of the docs the contributor has found themself on.

The utility of the `/dev` section in released doc versions is very low IMO, as most readers want the most up-to-date info. As mentioned in the PR description, an alternative would be to always redirect to the `development` version, but it probably makes sense to keep the older versions accessible for reference/historical reasons.

---

Hopefully this makes sense now, but let me know if not! I assume that the +1s from Pamphile and Ralf were to what I was intending, but perhaps not.

> At least for me the dev link works from a stable release, no? (ignore current version switcher issue, so from 1.11.3)

Yeah this is fine and orthogonal.

[^1]: this PR was motivated by a discussion on the Scientific Python Discord where Pamphile was helping Brigitta with build issues

**Comment by tylerjereddy** [comment]
Ah ok, I took the other meaning of "development docs," the clarification is very helpful, thanks.

Maybe it would be helpful to have a small sample diff to show how intrusive/minor the change would be. Then we could adopt a practice of adding the patch/note to future stable release process, but as for backporting maybe we could indeed get away with manually patching what's already out there (don't really like that, but far less work if it can be done with high degree of safety).

**Comment by lucascolley** [comment]
> Maybe it would be helpful to have a small sample diff to show how intrusive/minor the change would be.

Indeed, and that's what I'm not sure of either. Ideally, I think it would be good to have a note on every page in the `/dev` section.

My initial question was whether there would be some clever way to make a PR to `main` and just hide the notes from the `development` version, or whether we need the backporting / addition to release process.

**Comment by melissawm** [comment]
We talked about this at the community meeting today and I just wanted to mention one aspect.
 
If we consider the option of just redirecting people to the latest version whenever the Development pages are visited, it may be confusing for users. Consider a user navigating the (say) API docs for version 1.11.0

If they navigate to the development docs, they would be immediately redirected to the latest version of the Development pages. If they now try to navigate back to the API docs, they will still be in the latest version of the pages - and may lose that context if they are not paying attention. To mitigate this, one option would be to use a version warning banner, which is a fairly simple change in the conf.py enabled in later versions of the pydata-sphinx-theme (note that this would depend on #16660 being resolved). Here's how that looks like (the red banner at the very top of the page):

![Screenshot_20231227_175913](https://github.com/scipy/scipy/assets/3949932/28fa37b8-bc59-4c6f-b25c-7d8413f09dde)

The backporting option would not be as elegant, but eliminates this kind of problem.

**Comment by rgommers** [comment]
Older version of the dev section of the docs are regularly useful to me, so I'd like to keep them where they are. Aside from that: a prominent banner or note will be useful indeed, so +1 on that.

**Comment by lucascolley** [comment]
same issue in skimage: scikit-image/scikit-image#gh-7138

**Comment by rgommers** [comment]
> If we consider the option of just redirecting people to the latest version whenever the Development pages are visited, it may be confusing for users. Consider a user navigating the (say) API docs for version 1.11.0

This is indeed not nice. What seems worse is that there is no way to get rid of the banner. I am using the dev docs only typically, and have to stare at this prominent red bar every time I use the docs now. It really needs an `x` to be able to close it, and that should be remembered for the duration of the browsing session.

In the absence of that, the cure here seems worse than the disease.

**Comment by lucascolley** [comment]
> It really needs an x to be able to close it, and that should be remembered for the duration of the browsing session.

Would definitely be nice if the theme included this!

I'm neutral on the issue of the banner being too prominent (hard to weigh up the benefits of it being there vs what it looks like), but I definitely think that this is overall negative if the banner shows on the contributor section of the docs. If we can't get that working soon, I'd be in favour of a revert.

**Comment by melissawm** [comment]
Relevant issue: https://github.com/pydata/pydata-sphinx-theme/issues/1384

**Comment by lucascolley** [comment]
That sounds perfectly reasonable to me and a good idea, but what I really want is a banner that shows on the stable version of the contributor guide, pointing to the "latest" version. If we can do both of those that would be great,

**Comment by rgommers** [comment]
@melissawm proposal SGTM and can probably be done now. The added feature @lucascolley proposes for only the contributor/development section sounds good too, as long as it's easy to do (i.e., avoid custom fiddly JS code, that usually breaks sooner or later).

**Comment by lucascolley** [comment]
Now that gh-20184 is in I'll change the title to my remaining feature request
